File: ghidra/scripts/AnalyzeCriticalMaps.java
// Ghidra Jython script for analyzing discovered calibration pointers
// Uses Ghidra decompiler to determine TRUE dimensions and categories

import ghidra.app.decompiler.*;
import ghidra.program.model.listing.*;
import ghidra.program.model.address.*;
import ghidra.program.model.mem.*;
import ghidra.util.task.TaskMonitor;
import java.io.*;

public class AnalyzeCriticalMaps extends GhidraScript {

    // Critical addresses from sweep
    long[] CRITICAL_ADDRESSES = {
        0x040008, 0x04003A, 0x04005D, 0x042FD9,  // EWS
        0x191F6F, 0x1928F4, 0x19783C, 0x198140, 0x199000,  // Driver's Wish
        0x1B005D, 0x1B0158, 0x1B043A, 0x1B0497, 0x1B0FC5  // Torque
    };

    @Override
    protected void run() throws Exception {
        Program program = currentProgram;
        DecompilerInterface decompiler = state.getTool().getService(DecompileInterface);
        Listing listing = program.getListing();
        Memory memory = program.getMemory();
        AddressSpace space = program.getAddressFactory().getDefaultAddressSpace();

        println("Starting Ghidra analysis of critical calibration pointers...\n");

        for (long calib_addr : CRITICAL_ADDRESSES) {
            analyzeAddress(program, decompiler, listing, memory, space, calib_addr);
        }

        println("\nAnalysis complete.");
    }

    void analyzeAddress(Program program, DecompilerInterface decompiler, Listing listing, 
                        Memory memory, AddressSpace space, long calib_addr) throws Exception {
        
        println("=== Analyzing 0x" + Long.toHexString(calib_addr).toUpperCase() + " ===");
        
        // Find functions that reference this calibration address
        String func_name = findReferencingFunction(listing, calib_addr);
        
        // Get decompiled C code
        String decompiled_code = "";
        if (func_name != null && !func_name.isEmpty()) {
            println("Function: " + func_name);
            Function func = listing.getFunctionByName(func_name);
            if (func != null) {
                DecompileResults results = decompiler.decompileFunction(func, 60, monitor);
                if (results != null && results.decompileCompleted()) {
                    decompiled_code = results.getDecompiledFunction().getC();
                    
                    // Print first 1000 chars of decompiled code
                    String preview = decompiled_code.length() > 1000 
                        ? decompiled_code.substring(0, 1000) + "..."
                        : decompiled_code;
                    println("Decompiled preview:");
                    println(preview);
                } else {
                    println("Decompile failed: " + results.getErrorMessage());
                }
            }
        } else {
            println("No referencing function found");
        }

        // Read data at calibration address
        byte[] data = new byte[512];
        try {
            memory.getBytes(space.getAddress(calib_addr), data);
        } catch (Exception e) {
            println("Error reading memory: " + e.getMessage());
            return;
        }

        String first_bytes_hex = bytesToHex(data, 64);
        println("\nFirst 64 bytes: " + first_bytes_hex);

        // Analyze structure from C code
        String dims = analyzeDimensions(decompiled_code, data);
        println("Dimensions: " + dims);

        // Determine category
        String category = determineCategory(calib_addr, dims);
        println("Category: " + category);

        println("\n");
    }

    String findReferencingFunction(Listing listing, long target_addr) {
        for (Function func : listing.getFunctions(true)) {
            try {
                InstructionIterator iter = listing.getInstructions(func.getEntryPoint(), true);
                while (iter.hasNext()) {
                    Instruction instr = iter.next();
                    Address addr = instr.getAddress();
                    
                    // Check operands for our target address
                    for (int i = 0; i < instr.getNumOperands(); i++) {
                        Object[] ops = instr.getOpObjects(i);
                        for (Object op : ops) {
                            if (op instanceof Address) {
                                if (((Address) op).getOffset() == target_addr) {
                                    return func.getName();
                                }
                            } else if (op instanceof Scalar) {
                                if (((Scalar) op).getUnsignedValue() == target_addr) {
                                    return func.getName();
                                }
                            }
                        }
                    }
                }
            } catch (Exception e) {
                // Skip functions that fail
            }
        }
        return null;
    }

    String analyzeDimensions(String decompiled_code, byte[] data) {
        // Check for struct definitions in C code
        if (decompiled_code.contains("struct") || decompiled_code.contains("typedef")) {
            if (decompiled_code.contains("struct {") || decompiled_code.contains("struct\n{")) {
                return "struct";
            }
        }
        
        // Check for array loops with known sizes
        if (decompiled_code.contains("for (") || decompiled_code.contains("while (")) {
            // Look for loop patterns like "for (i = 0; i < 256; i++)"
            int idx = decompiled_code.indexOf("for (");
            if (idx >= 0) {
                String for_clause = decompiled_code.substring(idx, Math.min(idx + 100, decompiled_code.length()));
                // Try to extract the loop bound
                int gt_idx = for_clause.indexOf("<");
                if (gt_idx >= 0) {
                    int semicolon_idx = for_clause.indexOf(";", gt_idx);
                    if (semicolon_idx > 0) {
                        String bound_str = for_clause.substring(gt_idx + 1, semicolon_idx).trim();
                        try {
                            int bound = Integer.parseInt(bound_str);
                            return bound + "x1";
                        } catch (NumberFormatException e) {
                            // Try to extract number from string
                            String num = bound_str.replaceAll("\\D+", "");
                            if (!num.isEmpty()) {
                                return num + "x1";
                            }
                        }
                    }
                }
            }
        }
        
        // Fall back to raw data analysis
        int axis_count = data[0] & 0xFF;
        if (axis_count >= 2 && axis_count <= 32) {
            int rows = axis_count;
            int cols = data.length > rows + 1 ? (data[rows + 1] & 0xFF) : 1;
            return rows + "x" + cols;
        } else if (axis_count == 1 || axis_count == 0) {
            // Likely a scalar or small array
            return "scalar/array";
        }
        
        return "unknown";
    }

    String determineCategory(long addr, String dims) {
        if (addr >= 0x040000 && addr < 0x045000) {
            return "EWS/Immobilizer";
        } else if (addr >= 0x190000 && addr < 0x1A0000) {
            return "Driver's Wish";
        } else if (addr >= 0x1AF000 && addr < 0x1B3000) {
            return "Torque/Limiter";
        } else if (addr >= 0x189000 && addr < 0x18B000) {
            return "DTC/Error";
        }
        
        // Use dimensions to help categorize
        if (dims.contains("Nx2") || dims.contains("mask")) {
            return "DTC/Error";
        }
        
        return "Uncategorized";
    }

    String bytesToHex(byte[] data, int len) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < Math.min(len, data.length); i++) {
            sb.append(String.format("%02x", data[i] & 0xFF));
            if (i < Math.min(len, data.length) - 1) {
                sb.append(" ");
            }
        }
        return sb.toString();
    }
}

File: ghidra/scripts/AnalyzeCriticalMaps.py
// Ghidra Jython script for analyzing discovered calibration pointers
// Uses Ghidra decompiler to determine TRUE dimensions and categories

from ghidra.program.model.listing import *
from ghidra.program.model.address import *
from ghidra.app.decompiler import *
from java.io import FileWriter
import json
from collections import Counter

# Critical addresses from sweep (selected samples)
CRITICAL_ADDRESSES = [
    # EWS candidates
    0x040008, 0x04003A, 0x04005D, 0x042FD9,
    # Driver's Wish candidates  
    0x191F6F, 0x1928F4, 0x19783C, 0x198140, 0x199000,
    # Torque limiters
    0x1B005D, 0x1B0158, 0x1B043A, 0x1B0497, 0x1B0FC5,
]

program = currentProgram
decompiler = state.getTool().getService(DecompileInterface)
listing = program.getListing()
memory = program.getMemory()
space = program.getAddressFactory().getDefaultAddressSpace()

results = []

for calib_addr in CRITICAL_ADDRESSES:
    # Get function containing a reference to this address
    # First, find instructions that reference this address
    refs_to_addr = []
    
    # Search for references to this calibration address
    for func in listing.getFunctions(True):
        code_iter = func.getEntryPoint().getAddressSet()
        try:
            instr_iter = listing.getInstructions(func.getEntryPoint(), True)
            while instr_iter.hasNext():
                instr = instr_iter.next()
                addr = instr.getAddress()
                
                # Check if this instruction references our calibration address
                try:
                    for i in range(instr.getNumOperands()):
                        op_objects = instr.getOpObjects(i)
                        for op in op_objects:
                            if hasattr(op, 'getAddress'):
                                op_addr = op.getAddress()
                                if op_addr and op_addr.getOffset() == calib_addr:
                                    refs_to_addr.append((func.getName(), addr))
                            elif hasattr(op, 'getValue'):
                                # Scalar value
                                val = op.getValue()
                                if isinstance(val, int) and val == calib_addr:
                                    refs_to_addr.append((func.getName(), addr))
                except:
                    pass
        except:
            pass
    
    # Get decompiled C code for the function
    decompiled_code = ""
    func_name = "unknown"
    
    if refs_to_addr:
        func_name = refs_to_addr[0][0]
        func = listing.getFunctionByName(func_name)
        if func:
            try:
                results_decomp = decompiler.decompileFunction(func, 60, monitor)
                if results_decomp and results_decomp.decompileCompleted():
                    decompiled_code = results_decomp.getDecompiledFunction().getC()
            except:
                decompiled_code = "[decompile error]"
    
    # Read data at calibration address
    data = bytearray(512)
    try:
        memory.getBytes(space.getAddress(calib_addr), data)
    except:
        data = bytearray(512)
    
    # Analyze structure
    first_bytes_hex = " ".join(f"{b:02x}" for b in data[:64])
    
    # Check for array loop patterns in C code
    is_array_loop = False
    array_size = 0
    if "for (" in decompiled_code or "while (" in decompiled_code:
        is_array_loop = True
        # Look for common array iteration patterns
        import re
        match = re.search(r'for\s*\(\s*\w+\s*=\s*\w+;\s*\w+\s*<\s*(\d+)', decompiled_code)
        if match:
            array_size = int(match.group(1))
    
    # Determine dimensions based on C code analysis
    if "struct" in decompiled_code or "typedef" in decompiled_code:
        # Check for struct definitions
        if "struct {" in decompiled_code:
            # Multi-dimensional struct
            rows_match = re.search(r'(\d+)\s*\[\s*(\d+)\s*\]', decompiled_code)
            if rows_match:
                dims = f"{rows_match.group(1)}x{rows_match.group(2)}"
            else:
                dims = "struct"
        else:
            dims = "struct"
    elif is_array_loop and array_size > 0:
        # Single-dimensional array with known loop count
        dims = f"{array_size}x1"
    elif calib_addr >= 0x189000 and calib_addr < 0x18B000:
        # DTC mask range - typically small arrays
        dims = "Nx2 (DTC mask)"
    else:
        # Fall back to raw data analysis
        axis_count = data[0] & 0xFF
        if 2 <= axis_count <= 32:
            rows = axis_count
            cols = data[rows+1] & 0xFF if rows+1 < len(data) else 1
            dims = f"{rows}x{cols}"
        else:
            dims = "unknown"
    
    # Categorize
    category = "Uncategorized"
    if calib_addr >= 0x040000 and calib_addr < 0x045000:
        category = "EWS/Immobilizer"
    elif calib_addr >= 0x190000 and calib_addr < 0x1A0000:
        category = "Driver's Wish"
    elif calib_addr >= 0x1AF000 and calib_addr < 0x1B3000:
        category = "Torque/Limiter"
    elif calib_addr >= 0x189000 and calib_addr < 0x18B000:
        category = "DTC/Error"
    
    result = {
        "calib_addr": f"0x{calib_addr:06X}",
        "category": category,
        "dimensions": dims,
        "is_array_loop": is_array_loop,
        "array_size": array_size if is_array_loop else 0,
        "first_bytes_hex": first_bytes_hex,
        "refs_from_func": func_name,
        "decompiled_preview": decompiled_code[:500] if len(decompiled_code) > 500 else decompiled_code
    }
    
    results.append(result)
    print(f"Analyzed 0x{calib_addr:06X}: {category} | {dims} | refs: {func_name}")

# Save results
output_path = "/home/brandon/Documents/EDC17CP09/ghidra_analysis_results.json"
with open(output_path, 'w') as f:
    json.dump(results, f, indent=2)

print(f"\nSaved to {output_path}")

# Print summary
cat_counts = Counter(r['category'] for r in results)
print("\nCategory breakdown:")
for cat, count in cat_counts.most_common():
    print(f"  {cat}: {count}")
